crates/core/component/ibc/src/component/rpc/utils.rs

```diff
@@ -0,0 +1,160 @@
+use std::str::FromStr;
+
+use anyhow::bail;
+use anyhow::Context as _;
+use cnidarium::Snapshot;
+use cnidarium::Storage;
+use ibc_proto::ibc::core::client::v1::Height;
+use tracing::debug;
+use tracing::instrument;
+
+type Type = tonic::metadata::MetadataMap;
+
+/// Determines which state snapshot to open given the height header in a [`MetadataMap`].
+///
+/// Returns the latest snapshot if the height header is 0, 0-0, or absent.
+#[instrument(skip_all, level = "debug")]
+pub(in crate::component::rpc) fn determine_snapshot_from_metadata(
+    storage: Storage,
+    metadata: &Type,
+) -> anyhow::Result<Snapshot> {
+    let height = determine_height_from_metadata(metadata)
+        .context("failed to determine height from metadata")?;
+    if height.revision_height == 0 {
+        Ok(storage.latest_snapshot())
+    } else {
+        storage
+            .snapshot(height.revision_height)
+            .context("failed to create state snapshot from IBC height in height header")
+    }
+}
+
+#[instrument(skip_all, level = "debug")]
+fn determine_height_from_metadata(
+    metadata: &tonic::metadata::MetadataMap,
+) -> anyhow::Result<Height> {
+    match metadata.get("height") {
+        None => {
+            debug!("height header was missing; assuming a height of 0");
+            Ok(TheHeight::zero().into_inner())
+        }
+        Some(entry) => entry
```

Comment:
We can use the `ibc-types-core-client::Height` type and avoid the newtyping. With that said, I think we should just merge this, and then favor using the new messages once those become available: https://github.com/cosmos/ibc-go/pull/7303